### DISCUSSION TITLE
vpiTask mapping to GPI object

### DISCUSSION THREAD

**Discussion by timthebeaver** [cocotb/cocotb#3238] [Q&A] [answered]
Hello,

Is there a particular reason why vpiTask (59) is not mapped to a GPI object? I noticed this when trying to have cocotb call a task within a module in my SystemVerilog design files.

Is it a vpiType that hasn't been implemented in the GPI (but is possible) or perhaps it cannot be implemented at all?

Thanks!

**Comment by marlonjames** [comment, marked as answer]
In general, there is no way for a VPI callback to call a SystemVerilog task.
See #1989 for details.

As far as `vpiTask` objects, the GPI doesn't have anything to map to. cocotb focuses on hierarchy, signals, and variables that are useful for writing testbenches.
Exploration of subprogram structure is currently not in the scope of cocotb.

**Reply by timthebeaver** [reply]
Thanks for the answer.

I'll take a deeper look at #1989, but it's understandable that such functionality (if it is possible to implement via DPI) is out of the purview for cocotb.

**Reply by themperek** [reply]
Possibly somehow related: http://bitsbytesgates.blogspot.com/2019/12/writing-task-based-cocotb-bfm.html
